### ISSUE TITLE
[Phase 3] Cold-Start Handling with CF Embeddings

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/contextual-bandit-recommender#4]
## Summary

Implement collaborative filtering embeddings and warm-start logic to handle new users and items. This addresses the cold-start problem that affects all recommendation systems.

**Parent Issue**: #1
**Depends On**: #2 (Core Bandit Algorithm), #3 (Feature Engineering)

## System Context

```
src/coldstart/
├── __init__.py
├── cf_embeddings.py    ← TruncatedSVD matrix factorization
└── warmup.py           ← Warm-start logic for new entities
```

### Files to Create

| File | Lines (est.) | Description |
|------|--------------|-------------|
| `src/coldstart/cf_embeddings.py` | ~100 | TruncatedSVD for user/item embeddings |
| `src/coldstart/warmup.py` | ~80 | Initialize bandit arms for new entities |
| `tests/test_coldstart.py` | ~100 | Cold-start handling tests |

### CF Embeddings
- [ ] Create `CFEmbedder` class using sklearn TruncatedSVD
- [ ] Implement `fit(interactions)` from user-item matrix
- [ ] Implement `get_user_embedding(user_id)` method
- [ ] Implement `get_item_embedding(item_id)` method
- [ ] Handle unknown users/items with average embedding fallback
- [ ] Store ID-to-index mappings

### Warm-Start Handler
- [ ] Create `WarmStartHandler` class
- [ ] Implement `is_cold_start(user_id)` detection
- [ ] Implement `warm_start_user(user_id, similar_users)` method
- [ ] Implement `warm_start_item(item_id, similar_items)` method
- [ ] Integrate with bandit model initialization

### Testing
- [ ] Test embedding dimensions
- [ ] Test unknown user/item fallback
- [ ] Test warm-start improves cold-start performance

### CF Embedder

```python
# src/coldstart/cf_embeddings.py
import numpy as np
from numpy.typing import NDArray
from sklearn.decomposition import TruncatedSVD
from scipy.sparse import csr_matrix
from typing import Dict, Optional
import pandas as pd

class CFEmbedder:
    """
    Collaborative filtering embeddings via matrix factorization.

    Uses TruncatedSVD to decompose user-item interaction matrix
    into low-dimensional embeddings for users and items.
    """

    def __init__(self, embedding_dim: int = 32, random_state: int = 42):
        self.embedding_dim = embedding_dim
        self.svd = TruncatedSVD(n_components=embedding_dim, random_state=random_state)
        self.user_embeddings: Optional[NDArray[np.float64]] = None
        self.item_embeddings: Optional[NDArray[np.float64]] = None
        self.user_id_to_idx: Dict[str, int] = {}
        self.item_id_to_idx: Dict[str, int] = {}
        self._avg_user_embedding: Optional[NDArray[np.float64]] = None
        self._avg_item_embedding: Optional[NDArray[np.float64]] = None

    def fit(self, interactions: pd.DataFrame) -> 'CFEmbedder':
        """
        Fit embeddings from interaction data.

        Args:
            interactions: DataFrame with columns [user_id, item_id, rating]

        Returns:
            self for chaining
        """
        # Build ID mappings
        unique_users = interactions['user_id'].unique()
        unique_items = interactions['item_id'].unique()
        self.user_id_to_idx = {uid: i for i, uid in enumerate(unique_users)}
        self.item_id_to_idx = {iid: i for i, iid in enumerate(unique_items)}

        # Build sparse interaction matrix
        rows = [self.user_id_to_idx[u] for u in interactions['user_id']]
        cols = [self.item_id_to_idx[i] for i in interactions['item_id']]
        data = interactions['rating'].values

        n_users = len(unique_users)
        n_items = len(unique_items)
        matrix = csr_matrix((data, (rows, cols)), shape=(n_users, n_items))

        # Fit SVD: matrix ≈ U @ S @ V^T
        # User embeddings = U @ sqrt(S)
        # Item embeddings = V^T.T @ sqrt(S) = V @ sqrt(S)
        U = self.svd.fit_transform(matrix)
        S_sqrt = np.sqrt(self.svd.singular_values_)
        V = self.svd.components_.T

        self.user_embeddings = U * S_sqrt  # (n_users, embedding_dim)
        self.item_embeddings = V * S_sqrt  # (n_items, embedding_dim)

        # Compute average embeddings for fallback
        self._avg_user_embedding = np.mean(self.user_embeddings, axis=0)
        self._avg_item_embedding = np.mean(self.item_embeddings, axis=0)

        return self

    def get_user_embedding(self, user_id: str) -> NDArray[np.float64]:
        """Get embedding for user, or average if unknown."""
        if user_id in self.user_id_to_idx:
            return self.user_embeddings[self.user_id_to_idx[user_id]]
        return self._avg_user_embedding.copy()

    def get_item_embedding(self, item_id: str) -> NDArray[np.float64]:
        """Get embedding for item, or average if unknown."""
        if item_id in self.item_id_to_idx:
            return self.item_embeddings[self.item_id_to_idx[item_id]]
        return self._avg_item_embedding.copy()
```

### Warm-Start Handler

```python
# src/coldstart/warmup.py
import numpy as np
from numpy.typing import NDArray
from typing import List, Optional
from .cf_embeddings import CFEmbedder

class WarmStartHandler:
    """Handle cold-start for new users and items."""

    def __init__(self, cf_embedder: CFEmbedder, similarity_threshold: float = 0.5):
        self.cf_embedder = cf_embedder
        self.similarity_threshold = similarity_threshold
        self.known_users: set = set()
        self.known_items: set = set()

    def is_cold_start_user(self, user_id: str) -> bool:
        """Check if user is new (cold-start)."""
        return user_id not in self.known_users

    def register_user(self, user_id: str) -> None:
        """Mark user as no longer cold-start."""
        self.known_users.add(user_id)

    def find_similar_users(self, user_id: str, top_k: int = 5) -> List[str]:
        """Find most similar known users based on embeddings."""
        target_emb = self.cf_embedder.get_user_embedding(user_id)

        similarities = []
        for known_id in self.known_users:
            known_emb = self.cf_embedder.get_user_embedding(known_id)
            sim = np.dot(target_emb, known_emb) / (
                np.linalg.norm(target_emb) * np.linalg.norm(known_emb) + 1e-8
            )
            if sim > self.similarity_threshold:
                similarities.append((known_id, sim))

        similarities.sort(key=lambda x: x[1], reverse=True)
        return [uid for uid, _ in similarities[:top_k]]
```

## Technical Challenges

| Challenge | Solution |
|-----------|----------|
| Sparse interaction matrix | Use scipy.sparse for memory efficiency |
| New user has no interactions | Fall back to average embedding |
| No similar users found | Use population-level statistics |
| Incremental updates | Batch refit periodically (skip online updates) |

## Definition of Done

- [ ] `CFEmbedder.fit()` learns embeddings from interactions
- [ ] `get_user_embedding()` returns correct shape for known and unknown users
- [ ] `get_item_embedding()` returns correct shape for known and unknown items
- [ ] Warm-start users perform better than pure cold-start baseline
- [ ] All tests pass

## Verification Commands

```bash
# Run tests
pytest tests/test_coldstart.py -v

# Test embedding learning
python -c "
import pandas as pd
from src.coldstart import CFEmbedder

# Create sample interactions
interactions = pd.DataFrame({
    'user_id': ['u1', 'u1', 'u2', 'u2', 'u3'],
    'item_id': ['i1', 'i2', 'i2', 'i3', 'i1'],
    'rating': [1.0, 0.5, 1.0, 0.8, 0.6]
})

embedder = CFEmbedder(embedding_dim=8)
embedder.fit(interactions)

print(f'User embedding shape: {embedder.get_user_embedding(\"u1\").shape}')
print(f'Unknown user shape: {embedder.get_user_embedding(\"new_user\").shape}')
"
```

## Time Estimate

3-4 hours

**If running over**: Skip incremental updates, use batch retraining only
